### PR TITLE
fix: send blobs when running ipfs-http-client in the browser

### LINKED ISSUE #3138: Implementation bug in normaliseInput
I am pretty sure following lines would throw `TypeError: iterator is not iterable` because `yield *` expression expects iterable and not an iterator.  

https://github.com/ipfs/js-ipfs/blob/8cb8c73037e44894d756b70f344b3282463206f9/packages/ipfs-core-utils/src/files/normalise-input.js#L73-L82

https://github.com/ipfs/js-ipfs/blob/8cb8c73037e44894d756b70f344b3282463206f9/packages/ipfs-core-utils/src/files/normalise-input.js#L114-L122

https://github.com/ipfs/js-ipfs/blob/8cb8c73037e44894d756b70f344b3282463206f9/packages/ipfs-core-utils/src/files/normalise-input.js#L191-L199

It is also worrying that no tests seem to catch that.

### PR DESCRIPTION
**Pull request by achingbrain**
> Context: 
> - This PR aims to improve performance of `ipfs.add` in `ipfs-http-client` in browser context by addressing findings from https://github.com/ipfs/js-ipfs/issues/3029.  
> - Alternative approach can be found in https://github.com/ipfs/js-ipfs/pull/3151

To support streaming of native types with no buffering, normalise add input to blobs and upload using native FormData when the http client is run in the browser.

That is, if the user passes a blob to the http client in the browser leave it alone as enumerating blob contents cause the file data to be read.

Browser FormData objects do not allow you to specify headers for each multipart part which means we can't pass UnixFS metadata via the headers so we turn the metadata into a querystring and append it to the field name for each multipart part as a workaround.

Fixes #3138

BREAKING CHANGES:

- Removes the `mode`, `mtime` and `mtime-nsec` headers from multipart requests
- Passes `mode`, `mtime` and `mtime-nsec` as querystring parameters appended to the field name of multipart requests

### CODE REVIEW

**Comment by achingbrain** [comment]
cc @Gozala I think this is a bit simpler approach than https://github.com/ipfs/js-ipfs/pull/3151, the PR is less than a quarter of the size, uses more native browser features and means we don't have to maintain custom implementations of Blob, File and FormData.

We do change how we pass metadata to the server but so far the only implementation that supports metadata is js-IPFS so I think the impact will be low.

File: packages/ipfs-core-utils/src/files/normalise-input/utils.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    yield result.value
+  }
+}
+
+function blobToIt (blob) {
+  if (typeof blob.stream === 'function') {
+    // firefox < 69 does not support blob.stream()
+    return browserStreamToIt(blob.stream())
```
Comment (Gozala) [inline packages/ipfs-core-utils/src/files/normalise-input/utils.js:36]: You can turn blob to stream more simply by `new Response(blob).body`

**Comment by Gozala** [comment]
> cc @Gozala I think this is a simpler approach than #3151, the PR is less than a quarter of the size, uses more native browser features and means we don't have to maintain custom implementations of Blob, File and FormData.

@achingbrain I think there are some tradeoffs between two implementations that I would like to call out so they are considered.

- #3184 File / Blob polyfills were introduced in node because that prevents having two different code paths based on runtime.
  - In my experience that tends to be better for long term maintenance of the code base.
  - Two are less likely to diverge or exhibit behavioral differences.
- #3184 is indeed is larger in size, because:
   1. It adds ton of comments jsdoc style and inline explaining what code does or why (as someone new to codebase I really wish code was more commented because it's not always obvious what it does or why e.g. I'm still not sure what the feal with self executable async generators)
   2. It introduces ton of new tests to ensure that regressions aren't introduced.
   3. It uses custom FromData encoder (which contributes to the size as well), however that is because that is what was decided in our design review meeting.
- #3184 It correctly handles iterators (as per #3138).
- #3184 It opts most changed files into TS type-checking. _I know it's not exactly the goal of that patch, but it did uncover #3138 gradually added types really help in working with complex code base)

Ultimately it is your decision to make and I hope above notes will inform it.

File: packages/ipfs/src/http/utils/multipart-request-parser.js [hunk 2]
```diff
@@ -92,6 +79,21 @@ async function * parseEntry (stream, options) {
     }
 
     const disposition = parseDisposition(part.headers['content-disposition'])
+    const query = qs.parse(disposition.name.split('?').pop())
```
Comment (Gozala) [inline packages/ipfs/src/http/utils/multipart-request-parser.js:82]: is go implementation ok with added query params ? As in is it going to ignore them ?

File: packages/ipfs-core-utils/src/files/normalise-input/normalise-content.browser.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+const errCode = require('err-code')
+const { Blob } = require('ipfs-utils/src/globalthis')
+
+const {
+  isBytes,
+  isBloby
+} = require('./utils')
+
+function toBlob (input) {
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/normalise-content.browser.js:11]: This file is only pulled in by the http client in the browser, not by the ipfs module in the browser or node, so this only happens when sending content over the HTTP API to a remote node and not when running a full node in-browser.  The output of this module doesn't touch the unixfs importer.

File: packages/ipfs-core-utils/src/files/normalise-input/normalise-content.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+const errCode = require('err-code')
+const { Buffer } = require('buffer')
+const {
+  isBytes,
+  isBloby,
+  browserStreamToIt,
+  blobToIt
+} = require('./utils')
+
+function toAsyncIterable (input) {
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/normalise-content.js:12]: It's probably an oversight.  We can split the single/multiple codepaths in a future PR too, that should make the logic easier to follow.

File: packages/ipfs-core-utils/src/files/normalise-input/normalise-input.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+const errCode = require('err-code')
+const {
+  isBytes,
+  isBloby,
+  isFileObject,
+  browserStreamToIt
+} = require('./utils')
+
+module.exports = function normaliseInput (input, normaliseContent) {
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/normalise-input.js:11]: No good reason, other than changing the method signature didn't seem relevant to solving the problem, which is that we currently buffer blob contents in memory.

File: packages/ipfs-core-utils/src/files/normalise-input/normalise-input.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  // must give us something
+  if (input === null || input === undefined) {
+    throw errCode(new Error(`Unexpected input: ${input}`), 'ERR_UNEXPECTED_INPUT')
+  }
+
+  // String
+  if (typeof input === 'string' || input instanceof String) {
+    return (async function * () { // eslint-disable-line require-await
+      yield toFileObject(input, normaliseContent)
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/normalise-input.js:20]: We can split the single/multiple codepaths in a future PR, that should make the logic easier to follow.

File: packages/ipfs-core-utils/src/files/normalise-input/normalise-input.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    })()
+  }
+
+  // Iterable<?>
+  if (input[Symbol.iterator]) {
+    return (async function * () { // eslint-disable-line require-await
+      const iterator = input[Symbol.iterator]()
+      const first = iterator.next()
+      if (first.done) return iterator
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/normalise-input.js:37]: To handle `ipfs.add([])` which should create an empty file.

File: packages/ipfs/src/http/utils/multipart-request-parser.js (hunk 2 above)
Comment (achingbrain) [reply packages/ipfs/src/http/utils/multipart-request-parser.js:82]: I've added a test for this, it ignores them.

File: packages/ipfs-core-utils/src/files/normalise-input/index.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+ * ```
+ *
+ * Into:
+ *
+ * ```
+ * AsyncIterable<{ path, mode, mtime, content: AsyncIterable<Buffer> }>
+ * ```
+ *
+ * @param input Object
+ * @return AsyncInterable<{ path, content: AsyncIterable<Buffer> }>
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/index.js:47]: This module doesn't convert to `Blob`, only to `AsyncIterable<Buffer>`.  index.browser.js converts to `Blob`.

File: packages/ipfs-core-utils/src/files/normalise-input/normalise-input.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/normalise-input.js]: It does make the code much smaller and we can remove a bunch of artificially induced async. I'm sold, have make the change.

File: packages/ipfs-core-utils/src/files/normalise-input/utils.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  return readBlob(blob)
+}
+
+async function * readBlob (blob, options) {
+  options = options || {}
+
+  const reader = new FileReader()
+  const chunkSize = options.chunkSize || 1024 * 1024
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/utils.js]: Well, you need *some* kind of chunk size. This has been removed anyway.

File: packages/ipfs-core-utils/src/files/normalise-input/normalise-content.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    return (async function * () { // eslint-disable-line require-await
+      const iterator = input[Symbol.iterator]()
+      const first = iterator.next()
+      if (first.done) return iterator
+
+      // Iterable<Number>
+      if (Number.isInteger(first.value)) {
+        yield toBuffer(Array.from((function * () {
+          yield first.value
+          yield * iterator
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/normalise-content.js]: Should be fixed now.

File: packages/ipfs-core-utils/src/files/normalise-input/utils.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
+'use strict'
+
+const { Buffer } = require('buffer')
+const { Blob, FileReader } = require('ipfs-utils/src/globalthis')
+
+function isBytes (obj) {
+  return Buffer.isBuffer(obj) || ArrayBuffer.isView(obj) || obj instanceof ArrayBuffer
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/utils.js]: Fixed!

File: packages/ipfs-core-utils/src/files/normalise-input/utils.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+function isBloby (obj) {
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/ipfs-core-utils/src/files/normalise-input/utils.js]: I have no strong opinion on this, have changed it.

**Pull request merged**